React/testing.md

```diff
@@ -1 +1,299 @@
+
+To keep snapshot files smaller we can engage in **shallow rendering**.
+
+### Shallow Rendering
+
+Enzyme has a function called `shallow` which will render a component with placeholders for subcomponents.  So if we shallowly rendered `App`, it would render `<NewPetForm />` in place of the fully rendered html the `NewPetForm` component generates.  Using this technique you can limit snapshot changes to the component you are testing.
+
+The following two tests show deep and shallow rendering in snapshot tests.
+
+```javascript
+// src/App.js
+import React from 'react';
+import ReactDOM from 'react-dom';
+import { mount, shallow } from 'enzyme';
+import App from './App';
+
+describe('<App />', () => {
+
+  test('will match the last snapshot with deep rendering', () => {
+    const wrapper = mount(<App />);
+    expect(wrapper).toMatchSnapshot();
+  });
+
+  test('that it renders App with shallow rendering', () => {
+    const wrapper = shallow(<App />);
+    expect(wrapper).toMatchSnapshot();
+  });
+});
+```
+
+**Question** Run the above two tests.  What do you notice about the differences in the two Snapshots?
+
+**Question** If we have a shallow snapshot of `App`, how do we get alerted when `Pet` changes?
+
+**So why would you ever use `mount`?**
+
+Use `mount` when you need to test the interaction between a container and child component.  Otherwise for unit testing use `shallow` almost exclusively.  You should **avoid** using `mount` for snapshot testing.
+
+| Enzyme Function | Used for |  Description |
+| ------ | ------ | ----- |
+|  mount  |  Deep Rendering | Renders the entire component and subcomponent to test their interactions  |
+|  shallow  |  shallow rendering  | Used for tests on a single component in isolation.
+
+## Testing User Interaction
+
+With snapshot testing we have tested how React components **render** to the DOM.  Next we need to test how the componets respond to user interaction.
+
+```javascript
+// src/components/NewPetForm.test.js
+...
+test('when the user enters a name in the text field, the field is updated', () => {
+  // Arrange
+  // Shallow-mount the wrapper
+  const wrapper = shallow( <NewPetForm addPetCallback={() => {}} />);
+  // Find the input in the component
+  let nameField = wrapper.find('input[name="name"]');
+
+  // Act
+  // Trigger a 'change' event
+  nameField.simulate('change', {
+    target: {
+      name: 'name',
+      value: 'Bob',
+    },
+  });
+  // Force the component to update (setState is async)
+  wrapper.update();
+  // Find the updated component
+  nameField = wrapper.find('input[name="name"]');
+
+  // Assert
+  expect(nameField.getElement().props.value).toEqual('Bob');
```

Comment:
Is this `props` referring literally to the HTML properties? This would be good to clarify since `props` is really loaded